web-api/src/mind.rs

```diff
@@ -11,16 +11,253 @@
+            .into_iter()
+            .map(|document| {
+                document.id.try_into().map(|id| {
+                    let snippet = if document.snippet.is_empty() {
+                        document.title.to_string()
+                    } else {
+                        document.snippet.to_string()
+                    };
+                    let category = if document.category.is_empty() {
+                        if document.subcategory.is_empty() {
+                            None
+                        } else {
+                            Some(document.subcategory.to_string())
+                        }
+                    } else {
+                        Some(document.category.to_string())
+                    };
+                    let properties = [
+                        (
+                            CATEGORY_ID.clone(),
+                            DocumentProperty(Value::String(document.category)),
+                        ),
+                        (
+                            SUBCATEGORY_ID.clone(),
+                            DocumentProperty(Value::String(document.subcategory)),
+                        ),
+                        (
+                            TITLE_ID.clone(),
+                            DocumentProperty(Value::String(document.title)),
+                        ),
+                        (
+                            SNIPPET_ID.clone(),
+                            DocumentProperty(Value::String(document.snippet)),
+                        ),
+                        (
+                            URL_ID.clone(),
+                            DocumentProperty(Value::String(document.url)),
+                        ),
+                    ]
+                    .into();
+
+                    IngestedDocument {
+                        id,
+                        snippet,
+                        properties,
+                        category,
+                    }
+                })
+            })
+            .try_collect()?;
+
+        Ok(Json(Self { documents }))
+    }
+}
+
+impl UpdateInteractions {
+    fn new(documents: &[String]) -> Result<Json<Self>, Error> {
+        let documents = documents
+            .iter()
+            .map(|document| {
+                document.try_into().map(|document_id| UserInteractionData {
+                    document_id,
+                    interaction_type: UserInteractionType::Positive,
+                })
+            })
+            .try_collect()?;
+
+        Ok(Json(Self { documents }))
+    }
+}
+
+impl PersonalizedDocumentsQuery {
+    fn new(count: Option<usize>, documents: Option<&[String]>) -> Result<Query<Self>, Error> {
+        let documents = documents
+            .map(|documents| documents.iter().map(TryInto::try_into).try_collect())
+            .transpose()?;
+
+        Ok(Query(Self { count, documents }))
+    }
+}
+
+trait PersonalizedDocumentsResponder
+where
+    Self: Responder<Body = EitherBody<String>> + Sized,
```

Comment:
i tried to minimize the changes on the routes, which requires to have some cumbersome unpacking of the type-erased responses. the other option would be to split the handler logic from the routing code and run the tests directly on the handlers instead of the routes.